blog/content/post/2020-08-05-PCIe-features.md

```diff
@@ -0,0 +1,213 @@
+
+OS can manage the power environment of a Function directly by accessing registers
+corresponding to this capability. OS can set Device state to one of 4 states. 2
+of them (D0, D4) are mandatory and other 2 (D2, D3) can be optionally implemented.
+D0 state is a state where the Function is fully operational and uses full power,
+every next state (D1,D2,D3) uses lower power, but also takes more time to recover
+to D0.
+
+- **Message Signaled Interrupts**
+
+PCI has pins to let the central interrupt controller know that it needs service.
+This improves efficiency of the CPU which does not need to check every Function
+periodically, but it also significantly increases the number of needed lines. In
+PCIe instead of additional lines special messages are used to signal interrupt.
+Message Signaled Interrupts (MSI) allows the Function to write a small amount of
+interrupt-describing data to a special memory-mapped I/O address, and the interrupt
+controller then delivers the corresponding interrupt to a processor. Data and
+the address to which they are being written are located in MSI capability structure.
+
+___
+
+**All following capabilities are optional.**
+
+___
+
+- **EXtended Message Signaled Interrupts**
+
+Extends MSI from 32 possible vectors to 2048 by placing a table containing
+addresses and messages of each vector in RAM instead of Device registers.
+Pointer to the table and its size is saved in the capability structure.
+
+- **Dynamic Power Allocation**
+
+This capability extends power states by additional 32 states between D0 and D1.
+
+- **Power Budgeting**
+
+Its goal is to allocate power for PCIe hot plug Devices that are added to the
+system during runtime. This ensures that the system can allocate the proper
+amount of power and cooling for these Devices.
+
+- **Advanced Error Reporting**
+
+It provides additional registers that give error handling software more
+information to work with in diagnosing and recovering from problems.
+
+- **Virtual Channel**
+
+It adds buffers that act as queues for outgoing packets. This capability allows
+Device to assign one of 8 priorities to packets. Every priority has its own
+buffer. Packets from buffers with higher priorities are sent more frequently.
+
+- **TLP Processing Hints**
+
+It adds the possibility to use cache inside Switches and the root complex to
+alleviate the need for RAM usage. The idea is similar to the idea of ​​adding
+cache to processors; adding a small amount of very fast memory to save soon
+needed data is a more efficient approach than saving this data in bigger and
+slower RAM.
+
+- **Resizable BAR Capability**
+
+It replaces the BAR registers from HEADER with bigger registers in this
+capability structure, allowing bigger memory space allocation for Functions.
+
+- **Access Control Services**
+
+The PCIe specification allows for peer-to-peer transactions. This means that it
+is possible and even desirable in some cases for one PCIe Endpoint to send data
```

Comment:
The second sentence suggests that the fact that it is sometimes desirable is a result of the specification
```suggestion
The PCIe specification allows for peer-to-peer transactions. It
is possible and even desirable in some cases for one PCIe Endpoint to send data
```